components/greaselion/browser/greaselion_download_service.cc

```diff
@@ -185,21 +201,54 @@ GreaselionDownloadService::GreaselionDownloadService(
     is_dev_mode_ = true;
     resource_dir_ = forced_local_path;
     LoadDirectlyFromResourcePath();
+    dev_mode_task_runner_ = base::ThreadPool::CreateSequencedTaskRunner(
+        {base::TaskPriority::BEST_EFFORT,
+         base::TaskShutdownBehavior::CONTINUE_ON_SHUTDOWN, base::MayBlock()});
     dev_mode_path_watcher_ = std::make_unique<base::FilePathWatcher>();
-    if (!dev_mode_path_watcher_->Watch(resource_dir_,
-        true /*recursive*/,
-        base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged,
-            weak_factory_.GetWeakPtr()))) {
-      LOG(ERROR) << "Greaselion could not watch filesystem for changes"
-          << " at path " << resource_dir_.LossyDisplayName();
-    }
+
+    using Callback = base::RepeatingCallback<void(bool)>;
+    base::FilePathWatcher::Callback file_path_watcher_callback =
+        base::BindRepeating(
+            [](scoped_refptr<base::SequencedTaskRunner> main_sequence,
+               const Callback& callback, const base::FilePath&, bool error) {
+              main_sequence->PostTask(FROM_HERE,
+                                      base::BindOnce(callback, error));
+            },
+            base::SequencedTaskRunnerHandle::Get(),
+            base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged,
+                       weak_factory_.GetWeakPtr()));
+
+    // Start the watcher on a background sequence, reporting all events back to
+    // this sequence. base::Unretained is safe because the watcher instance
+    // lives on the target sequence and will be destroyed there in a subsequent
+    // task.
+    dev_mode_task_runner_->PostTaskAndReplyWithResult(
+        FROM_HERE,
+        base::BindOnce(&StartFilePathWatcher,
+                       base::Unretained(dev_mode_path_watcher_.get()),
+                       resource_dir_, base::FilePathWatcher::Type::kRecursive,
+                       file_path_watcher_callback),
+        base::BindOnce(
+            [](Callback callback, bool start_result) {
+              if (!start_result) {
+                callback.Run(/*error=*/true);
+              }
+            },
+            base::Bind(&GreaselionDownloadService::OnDevModeLocalFileChanged,
+                       weak_factory_.GetWeakPtr())));
   }
 #endif
   DETACH_FROM_SEQUENCE(sequence_checker_);
 }
 
-void GreaselionDownloadService::OnDevModeLocalFileChanged(
-    const base::FilePath& path, bool error) {
+GreaselionDownloadService::~GreaselionDownloadService() {
+  if (dev_mode_path_watcher_) {
+    dev_mode_task_runner_->DeleteSoon(FROM_HERE,
+                                      std::move(dev_mode_path_watcher_));
+  }
```

Comment:
Must be destroyed on same sequence that invokes `Watch`.

Reply:
It looks like there's a `reset()` on that `unique_ptr` somewhere else in the file - do we need to do the `DeleteSoon` there as well?

Reply:
Yeah, I see what you mean. We don't need to `reset` on error since it doesn't really accomplish anything, so I removed that and updated the log message.